ports/espressif/bindings/espnow/ESPNow.c

```diff
@@ -0,0 +1,943 @@
+        self->tx_packets = self->tx_responses;
+    }
+}
+
+void espnow_reset(void) {
+    espnow_deinit();
+    MP_STATE_PORT(espnow_singleton) = NULL;
+}
+
+STATIC mp_obj_t espnow_active(size_t n_args, const mp_obj_t *args) {
+    esp_espnow_obj_t *self = _get_singleton();
+    if (n_args > 1) {
+        if (mp_obj_is_true(args[1])) {
+            espnow_init();
+        } else {
+            espnow_deinit();
+        }
+    }
+    return self->recv_buffer != NULL ? mp_const_true : mp_const_false;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(espnow_active_obj, 1, 2, espnow_active);
+
+// ESPNow.config(['param'|param=value, ..])
+// Get or set configuration values. Supported config params:
+//    buffer: size of buffer for rx packets (default=514 bytes)
+//    timeout: Default read timeout (default=300,000 milliseconds)
+STATIC mp_obj_t espnow_config(
+    size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+
+    esp_espnow_obj_t *self = _get_singleton();
+    enum { ARG_get, ARG_buffer, ARG_timeout, ARG_rate };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_, MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_buffer, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = -1} },
+        { MP_QSTR_timeout, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = -1} },
+        { MP_QSTR_rate, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = -1} },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args,
+        MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    if (args[ARG_buffer].u_int >= 0) {
+        self->recv_buffer_size = args[ARG_buffer].u_int;
+    }
+    if (args[ARG_timeout].u_int >= 0) {
+        self->recv_timeout_ms = args[ARG_timeout].u_int;
+    }
+    if (args[ARG_rate].u_int >= 0) {
+        #if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(4, 3, 0)
+        _wifi_init();  // Call the wifi init code
+        check_esp_err(esp_wifi_config_espnow_rate(
+            ESP_IF_WIFI_STA, args[ARG_rate].u_int));
+        check_esp_err(esp_wifi_config_espnow_rate(
+            ESP_IF_WIFI_AP, args[ARG_rate].u_int));
+        #else
+        mp_raise_ValueError(MP_ERROR_TEXT("rate option not supported"));
+        #endif
+    }
+    if (args[ARG_get].u_obj == MP_OBJ_NULL) {
+        return mp_const_none;
+    }
+#define QS(x) (uintptr_t)MP_OBJ_NEW_QSTR(x)
+    // Return the value of the requested parameter
+    uintptr_t name = (uintptr_t)args[ARG_get].u_obj;
+    if (name == QS(MP_QSTR_buffer)) {
+        return mp_obj_new_int(self->recv_buffer_size);
+    } else if (name == QS(MP_QSTR_timeout)) {
+        return mp_obj_new_int(self->recv_timeout_ms);
+    } else {
+        mp_raise_ValueError(MP_ERROR_TEXT("unknown config param"));
+    }
+#undef QS
+
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_KW(espnow_config_obj, 1, espnow_config);
+
+// ESPNow.on_recv(recv_cb)
+// Set callback function to be invoked when a message is received.
```

Comment:
We don't have anything like callbacks like this elsewhere in CircuitPython. Instead we would poll using a `.message_received` or something. Is the callback called asynchronously when a message is received, like a MicroPython soft interrupt?

Reply:
> Is the callback called asynchronously when a message is received, like a MicroPython soft interrupt?

Yes it is called asynchronously. `on_recv` is removed now.